### ISSUE TITLE
Add no-mutable-operate

### ISSUE THREAD

**Issue by suguanYang** [eslint/eslint#11642]
**Please describe what the rule should do:**
Add a no-mutable-operations rule, in order to forbid operations like `var`, `pop`, `shift`, etc

**What category of rule is this? (place an "X" next to just one item)**

[ ] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[x] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
// error, no mutable operation
var ressignable = 1

const arr = [1, 2,3]
// // error, no mutable operation
arr.pop()

```

**Why should this rule be included in ESLint (instead of a plugin)?**
Eslint as a most popular framework in javascript ecosystem， it could push forward functional programming in javascript, and make a new age for javascript.

**Are you willing to submit a pull request to implement this rule?**
yes my pleasure

**Comment by platinumazure** [comment]
Hi @suguanYang, thanks for the issue.

The first part already exists as the [`no-var`](https://eslint.org/docs/rules/no-var) rule.

As for the second rule, that would be pretty difficult to do without type checking, and there would be no way to account for newer types anyway. So what I would recommend is using [`no-restricted-syntax`](https://eslint.org/docs/rules/no-restricted-syntax) to flag operations you're particularly worried about. For example, you could use the selector `CallExpression[callee.property.name='pop']` to flag all function calls on a property named `pop`.

If you can think of specific cases that `no-var` and `no-restricted-syntax` don't handle, that could be broadly applicable to JavaScript developers, we're happy to evaluate it. As it is, I think your current request is a little too broad, but other team members might disagree.

**Comment by suguanYang** [comment]
Hi @platinumazure, thanks for the reply.

I will think more about your suggestions and do some practice on it.

Thanks.

**Comment by g-plane** [comment]
For preventing mutable array, you can consider TypeScript and using `ReadonlyArray`.
